lms/djangoapps/grades/models.py

```diff
@@ -184,12 +163,29 @@ def bulk_read(cls, course_key):
         Arguments:
             course_key: The course identifier for the desired records
         """
-        prefetched = get_cache(cls.CACHE_NAMESPACE).get(cls._cache_key(course_key))
-        if not prefetched:
+        prefetched = get_cache(cls._CACHE_NAMESPACE).get(cls._cache_key(course_key), None)
+        if prefetched is None:
             prefetched = cls._initialize_cache(course_key)
         return prefetched
 
     @classmethod
+    def cached_get_or_create(cls, blocks):
+        prefetched = get_cache(cls._CACHE_NAMESPACE).get(cls._cache_key(blocks.course_key))
+        if prefetched is not None:
+            model = prefetched.get(blocks.hash_value)
+            if not model:
+                model = cls.objects.create(
+                    hashed=blocks.hash_value, blocks_json=blocks.json_value, course_id=blocks.course_key,
+                )
+                cls._update_cache(blocks.course_key, [model])
+        else:
+            model, _ = cls.objects.get_or_create(
+                hashed=blocks.hash_value,
+                defaults={u'blocks_json': blocks.json_value, u'course_id': blocks.course_key},
+            )
+        return model
```

Comment:
**Note to reviewers:** Adding a prefetched cache for `VisibleBlocks` was also needed to reduce the queries when updating the course with `force_update_subsections=True`.